tst/unit/kokkos_abstraction.cpp

```diff
@@ -33,238 +39,293 @@ using parthenon::ParArray3D;
-  std::uniform_real_distribution<Real> dis(-1.0, 1.0);
-
-  const int N = 32;
-  ParArray4D<Real> arr_dev("device", N, N, N, N);
-  auto arr_host_orig = Kokkos::create_mirror(arr_dev);
-  auto arr_host_mod = Kokkos::create_mirror(arr_dev);
-
-  // initialize random data on the host not using any wrapper
-  for (int n = 0; n < N; n++)
-    for (int k = 0; k < N; k++)
-      for (int j = 0; j < N; j++)
-        for (int i = 0; i < N; i++)
-          arr_host_orig(n, k, j, i) = dis(gen);
-
-  // Copy host array content to device
-  Kokkos::deep_copy(arr_dev, arr_host_orig);
-
-  // increment data on the device using prescribed wrapper
-  parthenon::par_for(
-      loop_pattern, "unit test 4D", exec_space, 0, N - 1, 0, N - 1, 0, N - 1, 0, N - 1,
-      KOKKOS_LAMBDA(const int n, const int k, const int j, const int i) {
-        arr_dev(n, k, j, i) += static_cast<Real>(i + N * (j + N * (k + n)));
-      });
-
-  // Copy array back from device to host
-  Kokkos::deep_copy(arr_host_mod, arr_dev);
-
-  bool all_same = true;
-
-  // compare data on the host
-  for (int n = 0; n < N; n++)
-    for (int k = 0; k < N; k++)
-      for (int j = 0; j < N; j++)
-        for (int i = 0; i < N; i++)
-          if (arr_host_orig(n, k, j, i) + static_cast<Real>(i + N * (j + N * (k + n))) !=
-              arr_host_mod(n, k, j, i)) {
-            all_same = false;
-          }
-
-  return all_same;
-}
-
-TEST_CASE("par_for loops", "[wrapper]") {
-  auto default_exec_space = DevExecSpace();
-
-  SECTION("1D loops") {
-    REQUIRE(test_wrapper_1d(parthenon::loop_pattern_flatrange_tag, default_exec_space) ==
-            true);
+      }
+    } else {
+      for (int j = 0; j < N; j++) {
+        indices[Rank - LoopsLeft] = j;
+        all_same = par_for_comp<LoopsLeft - 1>(Sequence<Rank - 1>());
+      }
+    }
+    return all_same;
   }
 
-  SECTION("2D loops") {
-    REQUIRE(test_wrapper_2d(parthenon::loop_pattern_mdrange_tag, default_exec_space) ==
-            true);
+  template <typename, lbounds, typename, typename>
+  struct dispatch {};
+
+  template <typename Pattern, lbounds bound_type, size_t... Ids, typename... Ts>
+  struct dispatch<Pattern, bound_type, std::index_sequence<Ids...>,
+                  parthenon::TypeList<Ts...>> {
+    template <typename view_t>
+    void execute(DevExecSpace exec_space, view_t &dev, int *int_bounds,
+                 parthenon::IndexRange *bounds) {
+      const auto functor = KOKKOS_CLASS_LAMBDA(Ts... args) {
+        dev(std::forward<decltype(args)>(args)...) +=
+            increment_data(std::forward<decltype(args)>(args)...);
+      };
+      if constexpr (bound_type == lbounds::integer) {
+        parthenon::par_for(Pattern(), "unit test ND integer bounds", exec_space,
+                           int_bounds[Ids]..., functor);
+      } else {
+        parthenon::par_for(Pattern(), "unit test ND IndexRange bounds", exec_space,
+                           bounds[Ids]..., functor);
+      }
```

Comment:
this was the culprit for the 
```
error( cudaErrorInvalidDeviceFunction)
```
errors. For whatever reason having the two `KOKKOS_CLASS_LAMBDA`s in the two `if constexpr` branches here was causing errors. Oddly enough I'm doing the same thing in the `par_reduce` tests and it works fine...